# SQL
Компания купила крупный сервис для чтения книг по подписке.

**Задача**-проанализировать базу данных.

В базеданных — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [ ]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

 ## Исследование таблиц

In [ ]:
# получаем сырые данные
query = '''
            SELECT * FROM books
            LIMIT 5
        '''
books=pd.io.sql.read_sql(query, con = engine)
books.head(5)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [ ]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           5 non-null      int64 
 1   author_id         5 non-null      int64 
 2   title             5 non-null      object
 3   num_pages         5 non-null      int64 
 4   publication_date  5 non-null      object
 5   publisher_id      5 non-null      int64 
dtypes: int64(4), object(2)
memory usage: 368.0+ bytes


In [ ]:
# получаем сырые данные
query = '''
            SELECT * FROM authors
            LIMIT 5
        '''
authors=pd.io.sql.read_sql(query, con = engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [ ]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  5 non-null      int64 
 1   author     5 non-null      object
dtypes: int64(1), object(1)
memory usage: 208.0+ bytes


In [ ]:
# получаем сырые данные
query = '''
            SELECT * FROM publishers
            LIMIT 5
        '''
publishers=pd.io.sql.read_sql(query, con = engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [ ]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  5 non-null      int64 
 1   publisher     5 non-null      object
dtypes: int64(1), object(1)
memory usage: 208.0+ bytes


In [ ]:
# получаем сырые данные
query = '''
            SELECT * FROM ratings
            LIMIT 5
        '''
ratings=pd.io.sql.read_sql(query, con = engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [ ]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  5 non-null      int64 
 1   book_id    5 non-null      int64 
 2   username   5 non-null      object
 3   rating     5 non-null      int64 
dtypes: int64(3), object(1)
memory usage: 288.0+ bytes


In [ ]:
# получаем сырые данные
query = '''
            SELECT * FROM reviews
            LIMIT 5
        '''
reviews=pd.io.sql.read_sql(query, con = engine)
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [ ]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  5 non-null      int64 
 1   book_id    5 non-null      int64 
 2   username   5 non-null      object
 3   text       5 non-null      object
dtypes: int64(2), object(2)
memory usage: 288.0+ bytes


## Сколько книг вышло после 1 января 2000 года

In [ ]:
query = '''
            SELECT COUNT(title) as count_of_books
            FROM books
            WHERE publication_date > '01-01-2000'
        '''
pd.io.sql.read_sql(query, con = engine)

count_of_books
0             819

После 1 января 2000 года вышло 819 книг

## Количество обзоров и средняя оценка для каждой книги

In [ ]:
query = ''' SELECT b.title AS book_title,
                   COUNT(DISTINCT re.review_id) AS count,
                   ROUND(AVG(ra.rating)) as avg_rating
            FROM books AS b
            LEFT JOIN ratings AS ra ON ra.book_id = b.book_id
            LEFT JOIN reviews AS re ON re.book_id = b.book_id
            GROUP BY b.title, b.book_id
            ORDER BY count DESC
        '''
pd.io.sql.read_sql(query, con = engine)

book_title  count  avg_rating
0                              Twilight (Twilight  #1)      7         4.0
1                                  Water for Elephants      6         4.0
2                                     The Glass Castle      6         4.0
3    Harry Potter and the Prisoner of Azkaban (Harr...      6         4.0
4    The Curious Incident of the Dog in the Night-Time      6         4.0
..                                                 ...    ...         ...
995    Anne Rice's The Vampire Lestat: A Graphic Novel      0         4.0
996                            The Natural Way to Draw      0         3.0
997   The Cat in the Hat and Other Dr. Seuss Favorites      0         5.0
998                          Essential Tales and Poems      0         4.0
999  Disney's Beauty and the Beast (A Little Golden...      0         4.0

[1000 rows x 3 columns]

Для книг написано разное количество обзоров, также варьируются средние оценки книг

## Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [ ]:
query = ''' SELECT publisher,
            COUNT(book_id) AS count_books
            FROM publishers AS p
            LEFT JOIN books AS b ON p.publisher_id = b.publisher_id
            WHERE num_pages > 50
            GROUP BY publisher
            ORDER BY count_books DESC
            LIMIT 1
        '''
pd.io.sql.read_sql(query, con = engine)

publisher  count_books
0  Penguin Books           42

Наибольшее число книг толще 50 страниц выпустило издательство Penguin Books

## Автор с самой высокой средней оценкой книг

In [ ]:
query = ''' SELECT a.author AS author,
                   ROUND(AVG(ratings),2) AS rating
            FROM (
                  SELECT book_id,
                         AVG(rating) AS ratings
                  FROM ratings
                  GROUP BY book_id
                  HAVING COUNT(ratings) > 50
                  ) AS ra
            JOIN  books AS b ON ra.book_id=b.book_id
            JOIN  authors AS a ON b.author_id=a.author_id
            GROUP BY author
            ORDER BY rating DESC
            LIMIT 1
        '''
pd.io.sql.read_sql(query, con = engine)

author  rating
0  J.K. Rowling/Mary GrandPré    4.28

Автором с самой высокой средней оценкой книг является J.K. Rowling/Mary GrandPré

## Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [ ]:
query = ''' SELECT AVG(count)
            FROM (SELECT COUNT(review_id)
                  FROM reviews AS re
                  LEFT JOIN (SELECT username, COUNT(rating_id)
                             FROM ratings
                             GROUP BY username) AS ra ON ra.username = re.username
                  WHERE count > 50
                  GROUP BY re.username) AS count
        '''
pd.io.sql.read_sql(query, con = engine)

avg
0  24.333333

Cреднее количество обзоров от пользователей, которые поставли больше 50 оценок составляет 24 обзора

## Итоговый вывод

В ходе проведенного анализа было выяснено:
- После 1 января 2000 года вышло 819 книг
- Самое большое количество обзоров написано для книги Twilight (Twilight #1), имеющей средний рейтинг 4.0, наименьшее количество обзоров было характерно для нескольких книг.
- Наибольшее число книг толще 50 страниц выпустило издательство Penguin Books
- Автором с самой высокой средней оценкой книг является J.K. Rowling/Mary GrandPré
- Cреднее количество обзоров от пользователей, которые поставли больше 50 оценок составляет 24 обзора